src/rgw/rgw_auth_keystone.cc

```diff
@@ -226,26 +241,103 @@ TokenEngine::authenticate(const DoutPrefixProvider* dpp,
     return result_t::grant(std::move(apl));
   }
 
```

Comment:
Let me know if you think this code should be moved somewhere else inside a helper function.